src/context.rs

```diff
@@ -1,26 +1,226 @@
-use crate::Actor;
-use std::marker::PhantomData;
+use crate::envelope::{AsyncNonReturningEnvelope, Envelope, SyncNonReturningEnvelope};
+use crate::{Actor, Address, AsyncHandler, Handler, Message};
+use crate::manager::ManagerMessage;
+#[cfg(any(doc, feature = "with-tokio-0_2", feature = "with-async_std-1"))]
+use {std::time::Duration, crate::AddressExt};
 
 /// `Context` is used to signal things to the [`ActorManager`](struct.ActorManager.html)'s
 /// management loop. Currently, it can be used to stop the actor ([`Context::stop`](struct.Context.html#method.stop)).
 pub struct Context<A: Actor> {
+    /// Whether the actor is running. It is changed by the `stop` method as a flag to the `ActorManager`
+    /// for it to call the `stopping` method on the actor
     pub(crate) running: bool,
-    phantom: PhantomData<A>, // TODO(weak_address)
+    /// The address kept by the context to allow for the `Context::address` method to work.
+    address: Address<A>,
+    /// Notifications that must be stored for immediate processing.
+    pub(crate) immediate_notifications: Vec<Box<dyn Envelope<Actor = A>>>,
 }
 
 impl<A: Actor> Context<A> {
-    pub(crate) fn new() -> Self {
+    pub(crate) fn new(address: Address<A>) -> Self {
         Context {
             running: true,
-            phantom: PhantomData,
+            address,
+            immediate_notifications: Vec::with_capacity(1),
         }
     }
 
-    /// Stop the actor as soon as it has finished processing current message. This will mean that it
-    /// will be dropped, and [`Actor::stopped`](trait.Actor.html#method.stopped) will be called.
-    /// Any subsequent attempts to send messages to this actor will return the
-    /// [`Disconnected`](struct.Disconnected.html) error.
+    /// Stop the actor as soon as it has finished processing current message. This will mean that the
+    /// [`Actor::stopping`](trait.Actor.html#method.stopping) method will be called.
+    /// If that returns [`KeepRunning::No`](enum.KeepRunning.html#variant.No), any subsequent attempts
+    /// to send messages to this actor will return the [`Disconnected`](struct.Disconnected.html) error.
     pub fn stop(&mut self) {
         self.running = false;
     }
+
+    /// Get an address to the current actor if the actor is still running.
+    pub fn address(&self) -> Option<Address<A>> {
```

Comment:
Can it ever be not running? Or is it possible to guarantee that the context always references a running actor? (I don't know if this is viable just wanna ask questions if they're maybe useful xP)

Reply:
Yea it can be not running in the `stopped` and `stopping` methods. Reminds me that I should only return none in stopped because at that point the actor is about to be dropped so any address WILL become invalid, whereas in stopping it can recover.